src/database.ts

```diff
@@ -1,389 +1,383 @@
-		return callback(id);
-	}
+  constructor(
+    private objectStore: IDBObjectStore,
+    private key: string,
+    private schema: StoreSchema,
+    private api?: CrudApi<T>
+  ) {}
+
+  /**
+   * push local transactions to remote and try to fetch updates from remote
+   */
+  public sync(): Promise<void> {
+    type F = WithFlag<T>;
+    return new Promise(async (done, _) => {
+      const isOnline = this.api && isOnlineSupport(this.api)
+        ? await this.api.isOnline()
+        : undefined;
+      if (isOnline === false) {
+        done();
+        return;
+      }
+      const localItems = new Promise<F[]>((resolve, reject) => {
+        this.prepareTransaction();
+        const req = this.objectStore.index('flag').getAll();
+        req.onsuccess = () => resolve(req.result as F[]);
+        req.onerror = () => reject();
+      });
+      const actions = (await localItems).map(
+        (item: F): Promise<unknown> => {
+          switch (item.flag) {
+            case 'D':
+              return this.getId(item, (id: T[keyof T]) =>
+                this.deleteRemote(id, item)
+              );
+            case 'C':
+              return this.createRemote(item);
+            case 'U':
+              return this.updateRemote(item);
+            default:
+              return Promise.resolve();
+          }
+        }
+      );
+      await Promise.all(actions);
+      if (this.api && (isOnline === undefined || isOnline)) {
+        await this.api
+          .getAll()
+          .then(fromRemote => this.updateLocalStorage(fromRemote))
+          .catch((err) => console.error(`${this.key} sync failed with api`, err));
+      }
+      done();
+    });
+  }
+
+  public get(id: T[keyof T]): Promise<T | undefined> {
+    return new Promise((resolve, reject) => {
+      this.prepareTransaction();
+      const request = this.objectStore.get(id as unknown as IndexedKey);
+      request.onerror = err => reject(err);
+      request.onsuccess = async () => {
+        const result = request.result;
+        if (!result && this.api) {
+          const isOnline = isOnlineSupport(this.api)
+            ? await this.api.isOnline()
+            : undefined;
+          if (isOnline === undefined || isOnline) {
+            await this.api.get(id)
+              .then((response) => resolve(response))
+              .catch((e) => reject(e));
+            return;
+          }
+        }
+        resolve(!result || isDeleted(result) ? undefined : (result as T));
+      };
+    });
+  }
+
+  public getAll(): Promise<T[]> {
+    return new Promise<T[]>((resolve, reject) => {
+      this.prepareTransaction();
+      const request = this.objectStore.getAll();
+      request.onsuccess = async () => {
+        const result = request.result;
+        if (result?.length < 1 && this.api) {
+          const isOnline = isOnlineSupport(this.api)
+            ? await this.api.isOnline()
+            : undefined;
+          if (isOnline === undefined || isOnline) {
```

Comment:
```suggestion
          const isOnline = isOnlineSupport(this.api)
            ? await this.api.isOnline()
            : true;
          if (isOnline) {
```

I suggest to change this everywhere accordingly if it's never a different behavior.